File: checkers/static/js/ws_events.js
var socket = io();

function makeMove(from_square, to_square) {
    socket.emit('make_move', {from_cords: from_square, to_cords: to_square})
    currentPlayer = currentPlayer * -1;
    socket.emit('get_matrix');

    socket.on('receive_matrix', function(matrix) {
        drawChessBoard(matrix.matrix);
    });

    findPawnsToMove(currentPlayer)
        .then((moves) => {
            console.log(moves.pawns_cords);
            moves.pawns_cords.forEach(coord => {
                let data_cords = coord.join('-');
                let elements = document.querySelectorAll(`[data-position="${data_cords}"]`);

                highlightPossibleMoves(elements);
                enablePawnsToMove(elements);
            });
        })
        .catch((error) => {
            console.error(error);
        });
}

function findPawnsToMove(curr_player) {
    return new Promise((resolve, reject) => {
        socket.emit("find_pawns_to_move", {'current_player': curr_player });

        socket.on('found_legal_pawns', (data) => {
            let possible_moves = data;
            resolve(possible_moves);
        });
    });
}

function findPawnMoves(square) {
    return new Promise((resolve, reject) => {
        let pawn_cords = getPawnCords(square);

        socket.emit("find_pawn_moves", {
            'current_player': currentPlayer,
            'pawn': pawn_cords
        });

        socket.on('found_pawn_moves', (data) => {
            let possible_moves = data;
            resolve(possible_moves);
        });
    });
}

File: checkers/ws_events.py
from flask_socketio import emit, join_room, send

from checkers.checkers_game import Board
from checkers.websocket import socketio
from checkers_engine.move_validator import Validator

board = Board()
engine = Validator(board.matrix)


@socketio.on("get_matrix")
def send_matrix():
    emit("receive_matrix", {"matrix": engine.matrix})


@socketio.on("find_pawns_to_move")
def find_pawns_to_move(data):
    pawns_to_move = engine.find_pawns_available_to_move(data["current_player"])
    emit("found_legal_pawns", {"pawns_cords": pawns_to_move})


@socketio.on("find_pawn_moves")
def find_pawn_moves(data):
    pawn_legal_moves = engine.find_pawn_legal_moves(player=data["current_player"], pawn_cords=data["pawn"])

    emit("found_pawn_moves", {"pawns_cords": pawn_legal_moves})


@socketio.on("make_move")
def move_pawn(data):
    from_row, from_col = map(lambda x: int(x), (data["from_cords"].split("-")))
    to_row, to_col = map(lambda x: int(x), (data["to_cords"].split("-")))

    engine.matrix[from_row][from_col], engine.matrix[to_row][to_col] = (
        engine.matrix[to_row][to_col],
        engine.matrix[from_row][from_col],
    )
